java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java

```diff
@@ -0,0 +1,920 @@
+              + sparseTensorType);
+    }
+  }
+
+  /**
+   * Gets a copy of the data buffer.
+   *
+   * <p>As with {@link OnnxTensor} fp16 values are upcast into fp32 and returned as a {@link
+   * FloatBuffer}.
+   *
+   * @return The data buffer.
+   */
+  public Buffer getValuesBuffer() {
+    ByteBuffer buffer =
+        getValuesBuffer(OnnxRuntime.ortApiHandle, nativeHandle).order(ByteOrder.nativeOrder());
+    switch (info.type) {
+      case FLOAT:
+        if (info.onnxType == TensorInfo.OnnxTensorType.ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16) {
+          ShortBuffer shortBuffer = buffer.asShortBuffer();
+          int bufferCap = shortBuffer.capacity();
+          FloatBuffer output = FloatBuffer.allocate(bufferCap);
+          for (int i = 0; i < bufferCap; i++) {
+            output.put(fp16ToFloat(shortBuffer.get(i)));
+          }
+          output.rewind();
+          return output;
+        } else if (info.onnxType
+            == TensorInfo.OnnxTensorType.ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16) {
+          throw new IllegalArgumentException("BFloat16 is not supported.");
+        } else {
+          // regular fp32
+          FloatBuffer floatBuf = buffer.asFloatBuffer();
+          FloatBuffer output = FloatBuffer.allocate(floatBuf.capacity());
+          output.put(floatBuf);
+          output.rewind();
+          return output;
+        }
+      case DOUBLE:
+        {
+          DoubleBuffer doubleBuf = buffer.asDoubleBuffer();
+          DoubleBuffer output = DoubleBuffer.allocate(doubleBuf.capacity());
+          output.put(doubleBuf);
+          output.rewind();
+          return output;
+        }
+      case INT16:
+        {
+          ShortBuffer shortBuf = buffer.asShortBuffer();
+          ShortBuffer output = ShortBuffer.allocate(shortBuf.capacity());
+          output.put(shortBuf);
+          output.rewind();
+          return output;
+        }
+      case INT32:
+        {
+          IntBuffer intBuf = buffer.asIntBuffer();
+          IntBuffer output = IntBuffer.allocate(intBuf.capacity());
+          output.put(intBuf);
+          output.rewind();
+          return output;
+        }
+      case INT64:
+        {
+          LongBuffer longBuf = buffer.asLongBuffer();
+          LongBuffer output = LongBuffer.allocate(longBuf.capacity());
+          output.put(longBuf);
+          output.rewind();
+          return output;
+        }
+      case BOOL:
+      case INT8:
+      case UINT8:
+        {
+          ByteBuffer output = ByteBuffer.allocate(buffer.capacity());
+          output.put(buffer);
+          output.rewind();
+          return output;
+        }
+      case STRING:
+        throw new IllegalStateException("Unsupported data type String");
+      case UNKNOWN:
+      default:
+        throw new IllegalStateException("Unsupported data type");
+    }
+  }
+
+  /**
+   * Gets the shape of the (outer) indices.
```

Comment:
> Gets the shape of the (outer) indices. [](http://example.com/codeflow?start=4&length=39)

The name of the method does not match its description

Reply:
In the case of block sparse then this is `[num_blocks, block_shape1, block_shape2,...]`, in the case of coo this is `[numNonZero, dimension]` or `[numNonZero]`, and in the case of CSRC it's `[numRows+1]`.  I'm not sure why this requires returning an array of arrays. In the case of the block sparse it is packed, but that's how it's stored in the native code.

I agree if there are more sparse formats then it will need some more work, but as I don't know what the other possible sparse formats are then I'm not sure how to make it more flexible.